### PR TITLE
Added a caaspctl tool for the admin node

### PR DESCRIPTION
**Pull request by inercia**
Depends on #146

### CODE REVIEW

File: caasp-kvm/resources/scripts/caaspctl [hunk 1]
```diff
@@ -0,0 +1,1229 @@
...
+
+        logs)
+            $CAASPCTL logs velum
+            ;;
+
+        minions)
+            exec_velum_salt_fun "minions"
+            ;;
+
+        peniding|pending-minions)
```
Comment (MaximilianMeister) [inline caasp-kvm/resources/scripts/caaspctl]: pending

File: caasp-kvm/resources/scripts/caaspctl [hunk 2]
```diff
@@ -0,0 +1,1229 @@
...
+    orchestrate|orch   run a Salt orchestration (default: $ORCH_DEFAULT)
+    pillar             operate on the Salt pillar
+    grains             operate on the Salt grains
+    keys               operate on the Salt keys
+
+Kubernetes/Docker:
+
+    k8s                kubernetes-specific commands
+    registry|reg       create/manage a local Docker registry
+    cont-exec          run a command in a container
```
Comment (MaximilianMeister) [inline caasp-kvm/resources/scripts/caaspctl]: missing shortcuts cont-run|exec - maybe only exec is enough

File: caasp-kvm/resources/scripts/caaspctl [hunk 3]
```diff
@@ -0,0 +1,1229 @@
...
+
+EOF
+}
+
+k8s_command() {
+    local command=$1
+    shift
+    case "$command" in
+        kubeconfig)
+            k8s_kubeconfig $@
```
Comment (MaximilianMeister) [inline caasp-kvm/resources/scripts/caaspctl]: can't we just get this from the velum container? also i think the function doesn't need any args

and do you think it makes sense to add a kubectl command as well?

**Comment by ajaeger** [comment]
We already have caasp-cli - do we really need two tools? Can we merge them?

**Comment by ajaeger** [comment]
If we need two tools, should we use same naming scheme for both like caasp-ctl and caasp-cli?

**Comment by MaximilianMeister** [comment]
> If we need two tools, should we use same naming scheme for both like caasp-ctl and caasp-cli?

i'd prefer to call the user facing cli `caaspctl` (like all other tools, kubectl, etcdctl, systemctl ...)

i have opened https://github.com/kubic-project/caasp-cli/issues/7 let's see what comes out

**Comment by inercia** [comment]
> We already have caasp-cli - do we really need two tools? Can we merge them?

@ajaeger it would be difficult to merge both, as this `caaspctl` tool is a bash script while `caasp-cli` is a Go application. Besides that, `caasp-cli` would be used by regular users while `caaspctl` is more like an administrator or advanced users tool (in fact, it is only here in the CaaSP developement environment repo) 

> If we need two tools, should we use same naming scheme for both like caasp-ctl and caasp-cli?

As @MaximilianMeister said, I prefer the `<something>ctl` scheme as it is already present in many Linux tools...

File: caasp-kvm/resources/scripts/caaspctl [hunk 4]
```diff
@@ -0,0 +1,1229 @@
...
+                for x in $@ ; do
+                    do_add_pillar "${x%=*}" "${x#*=}"
+                done
+            else
+                log "No pillar values to set"
+            fi
+            ;;
+        ls|list)
+            log "Listing pillars (in the 'ca')"
+            salt_command 'ca' pillar.ls
```
Comment (ereslibre) [inline caasp-kvm/resources/scripts/caaspctl]: For getting this information I'd go with the salt runner: `salt-run pillar.show_pillar` to get a better overview.

File: caasp-kvm/resources/scripts/caaspctl [hunk 5]
```diff
@@ -0,0 +1,1229 @@
...
+            ;;
+        items)
+            log "Listing pillars (in the 'ca')"
+            salt_command 'ca' pillar.items
```
Comment (ereslibre) [inline caasp-kvm/resources/scripts/caaspctl]: Same here.

File: caasp-kvm/resources/scripts/caaspctl [hunk 6]
```diff
@@ -0,0 +1,1229 @@
...
+    shift
+    case "$command" in
+        set)
+            set -f
+            log "Setting grain $2=$3 in $1"
+            salt_command $1 grains.set "$2" "$3"
+            ;;
+        ls|list)
+            log "Listing grains (in the 'ca')"
+            salt_command 'ca' grains.ls
```
Comment (ereslibre) [inline caasp-kvm/resources/scripts/caaspctl]: Why not also use `$1` so we can list the grains on any minion?

File: caasp-kvm/resources/scripts/caaspctl [hunk 7]
```diff
@@ -0,0 +1,1229 @@
...
+            ;;
+        items)
+            log "Listing grains (in the 'ca')"
+            salt_command 'ca' grains.items
```
Comment (ereslibre) [inline caasp-kvm/resources/scripts/caaspctl]: Same here.

**Comment by ereslibre** [comment]
> As @MaximilianMeister said, I prefer the <something>ctl scheme as it is already present in many Linux tools...

I also agree with @inercia and @MaximilianMeister here.

File: caasp-kvm/resources/scripts/caaspctl [hunk 8]
```diff
@@ -0,0 +1,1229 @@
...
+VELUM_DB="velum_production"
+
+# the file where the Salt master hostname/IP is configured
+SALT_MASTER_CONFIG_FILE=/etc/salt/minion.d/master.conf
+
+# default port for the API server
+API_SERVER_PORT=6443
+
+# the Rail executable
+RAILS_EXE="/srv/velum/vendor/bundle/ruby/2.1.0/bin/rails"
```
[the remark this answers is not in the document]
Comment (inercia) [reply caasp-kvm/resources/scripts/caaspctl]: It does not seem to work: it is not in the path (at least not when doing `docker exec`).

File: caasp-kvm/resources/scripts/caaspctl [hunk 9]
```diff
@@ -0,0 +1,1229 @@
...
+reg_dump_catalog() {
+    log "Dumping registry catalog"
+    CURL_ARGS=
+    if [ -f $CA_CRT ] ; then
+        CURL_ARGS="$CURL_ARGS --cacert $CA_CRT"
+        CATALOG_URL="https://$REGISTRY/v2/_catalog"
+    else
+        CATALOG_URL="http://$REGISTRY/v2/_catalog"
+    fi
+    curl $CURL_ARGS -X GET "$CATALOG_URL"
```
[the remark this answers is not in the document]
Comment (inercia) [reply caasp-kvm/resources/scripts/caaspctl]: Right.

File: caasp-kvm/resources/scripts/caaspctl [hunk 10]
```diff
@@ -0,0 +1,1229 @@
...
+              -v $REGISTRY_CERTS_DIR:/certs \
+              -e REGISTRY_HTTP_TLS_CERTIFICATE=/certs/$(basename "$REG_CRT") \
+              -e REGISTRY_HTTP_TLS_KEY=/certs/$(basename "$REG_KEY") \
+              registry:2
+            ;;
+
+        stop )
+            log "Stopping the registry"
+            docker stop registry
+            docker rm registry
```
[the remark this answers is not in the document]
Comment (inercia) [reply caasp-kvm/resources/scripts/caaspctl]: It is not exactly the same.

File: caasp-kvm/resources/scripts/caaspctl (hunk 6 above)
Comment (inercia) [reply caasp-kvm/resources/scripts/caaspctl]: Ok, done.

File: caasp-kvm/resources/scripts/caaspctl (hunk 3 above)
Comment (inercia) [reply caasp-kvm/resources/scripts/caaspctl]: I don't know if the Velum container holds a `kubeconfig`, but this one is ready to be used (no need to use the `caasp-cli` afterwards).

File: caasp-kvm/resources/scripts/caaspctl (hunk 1 above)
Comment (inercia) [reply caasp-kvm/resources/scripts/caaspctl]: Fixed

File: caasp-kvm/resources/scripts/caaspctl [hunk 11]
```diff
@@ -0,0 +1,1229 @@
...
+                exec_velum_salt_fun "accept_minion(minion_id: '$i')"
+            done
+            ;;
+
+        fun|function|f|call|run)
+            exec_velum_salt_fun $@
+            ;;
+
+        enter)
+            log "Entering Valum container..."
```
[the remark this answers is not in the document]
Comment (inercia) [reply caasp-kvm/resources/scripts/caaspctl]: Fixed

File: caasp-kvm/resources/scripts/caaspctl (hunk 2 above)
Comment (inercia) [reply caasp-kvm/resources/scripts/caaspctl]: Ok

File: caasp-kvm/resources/scripts/caaspctl [hunk 12]
```diff
@@ -0,0 +1,1229 @@
...
+########################
+# Velum
+########################
+
+exec_velum_salt_fun() {
+    local fun=$1
+    log "Salt.$fun:"
+    exec_in_container "$CONTAINER_VELUM" \
+        entrypoint.sh bash -c \
+            "$RAILS_EXE runner \"require 'velum/salt'; require 'pp' ; pp(Velum::Salt.$fun)\""
```
[the remark this answers is not in the document]
Comment (inercia) [reply caasp-kvm/resources/scripts/caaspctl]: Ditto

**Comment by robdaemon** [comment]
I've added my comments on the scope of renaming the CaaSP CLI to that issue. In general, I'm a :-1: due to the timing of it, and the scope of the change. We're coming up towards the end of the 2.0 cycle, we should be focusing on bugs and missing features.

**Comment by kiall** [comment]
There is a lot of stuff here, some of it may be useful for end users, some of it only useful for devs - can we leave this for a while (i.e. until 2.0 is calmed down), and start a discussion around making some of this into something operator facing?

File: caasp-kvm/resources/scripts/caaspctl [hunk 13]
```diff
@@ -0,0 +1,1249 @@
...
+
+        * )
+            keys_usage
+            abort "Unknown keys command $command"
+            ;;
+    esac
+}
+
+########################
+# Salt keys
```
Comment (MaximilianMeister) [inline caasp-kvm/resources/scripts/caaspctl]: s/keys/logs

**Comment by MalloZup** [comment]
i would vote for closing this PR since is now deprecated.

What do you think? @mjura @inercia @MaximilianMeister 

https://github.com/kubic-project/caaspctl

**Comment by inercia** [comment]
> i would vote for closing this PR since is now deprecated.
> 
> What do you think? @mjura @inercia @MaximilianMeister
> 
> https://github.com/kubic-project/caaspctl

Yes, I think we can close it...

**Comment by MalloZup** [comment]
@inercia feel free to close it at moment i can not :cupid:

**Pull request closed without merging**